src/coreclr/jit/compiler.cpp

```diff
@@ -5431,6 +5431,13 @@ PhaseStatus Compiler::placeLoopAlignInstructions()
                 bbHavingAlign = prev;
                 JITDUMP("Marking " FMT_BB " before the loop with BBF_HAS_ALIGN for loop at " FMT_BB "\n", prev->bbNum,
                         block->bbNum);
+
+                // bbHavingAlign is in a loop, and precedes a nested loop
+                if (blockToLoop->GetLoop(bbHavingAlign) != nullptr)
+                {
+                    // If bbHavingAlign is a removable jump, it will be removed despite it having alignment
+                    bbHavingAlign->SetFlags(BBF_JMP_TO_NESTED_LOOP);
+                }
```

Comment:
I don't think I understand what we're trying to check here, but it doesn't seem to check for any form of nestedness or relationship between the two loops.

Reply:
I'm trying to check for examples like the one Kunal posted [here](https://github.com/dotnet/runtime/pull/97011#issuecomment-1893908462), such that `bbHavingAlign` is in a loop, and precedes the start of a nested loop. I'm assuming checking if `bbHavingAlign` is part of a loop is sufficient?

Reply:
>  I'm assuming checking if bbHavingAlign is part of a loop is sufficient?

Yes

Reply:
Well, that case should definitely be caught but `blockToLoop->GetLoop(bbHavingAlign)` could potentially not be a parent loop (it might just be a block jumping backwards to a separate loop, or skipping over this loop, or exiting a previous loop into this loop). But if the worry is about affecting other loops by starting to execute alignment in them then it seems like it's orthogonal to whether or not they are nested loops. Regardless, even if we are worried about it, wouldn't the previous behavior (before `BBJ_NONE` removal) have been to execute their alignment?

Reply:
> Regardless, even if we are worried about it, wouldn't the previous behavior (before `BBJ_NONE` removal) have been to execute their alignment?

true, there could be occurrences where we do execute `align` instructions in some other loop. What we are trying to see is if we are increasing those occurrences. Once the latest diffs come out, I am happy to chat offline to discuss how to proceed with this change.

Reply:
> Well, that case should definitely be caught but blockToLoop->GetLoop(bbHavingAlign) could potentially not be a parent loop

I see, so we should also check that `bbHavingAlign` is a jump to the next block (which is the start of the loop we're aligning), right?

Reply:
> > Well, that case should definitely be caught but blockToLoop->GetLoop(bbHavingAlign) could potentially not be a parent loop
> 
> I see, so we should also check that `bbHavingAlign` is a jump to the next block (which is the start of the loop we're aligning), right?

I think the current check is fine and we aren't really interested in nested loops vs top level loops -- just whether or not we are causing alignment to be executed in some other loop, which is what you're checking.

To answer your question regardless, you could check it using the loop tree (`loop->GetParent() == blockToLoop->GetLoop(bbHavingAlign)` for the immediate parent). Checking `BBJ_ALWAYS` would definitely make the nestedness relationship exist, but it wouldn't be a complete check (e.g. the previous block could also be a `BBJ_COND/BBJ_SWITCH` that skips the nested loops in some cases.